src/pip/_internal/commands/cache.py

```diff
@@ -207,22 +209,55 @@ def _find_http_files(self, options: Values) -> List[str]:
     def _find_wheels(self, options: Values, pattern: str) -> List[str]:
         wheel_dir = self._cache_dir(options, "wheels")
 
-        # The wheel filename format, as specified in PEP 427, is:
+        # The wheel filename format, as originally specified in PEP 427, is:
         #     {distribution}-{version}(-{build})?-{python}-{abi}-{platform}.whl
         #
         # Additionally, non-alphanumeric values in the distribution are
         # normalized to underscores (_), meaning hyphens can never occur
         # before `-{version}`.
```

Comment:
Unfortunately, due to bugs in build backends (like https://github.com/pypa/setuptools/issues/3777), this isn't guaranteed. Dashes are always normalized (as the wheel filename would be invalid otherwise) but the name may contain periods. For example, [`zope.sqlalchemy-3.1-py3-none-any.whl`](https://pypi.org/project/zope.sqlalchemy/3.1/#files). 

Normalization of the wheel filenames is going to be necessary to make this more robust. That can be saved for later, of course.